### PR TITLE
Release HTTP Request Body Earlier

### PR DESCRIPTION
**Pull request by original-brownbear**
We don't need to hold on to the request body past the beginning of sending
the response. There is no need to keep a reference to it until after the response
has been sent fully and we can eagerly release it here.
Note, this can be optimized further to release the contents even earlier but for now
this is an easy increment to saving some memory on the IO pool.

### CODE REVIEW

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpResponse.java [hunk 1]
```diff
@@ -45,8 +47,8 @@ public boolean containsHeader(String name) {
         return headers().contains(name);
     }
 
-    public Netty4HttpRequest getRequest() {
-        return request;
+    public HttpHeaders requestHeaders() {
+        return requestHeaders;
     }
```
Comment (keypointt) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpResponse.java:52]: sorry maybe a naive comment, is it better to name public api `getRequestHeaders` ?

**Review by keypointt — COMMENTED** [review]
minor non-blocking

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpResponse.java [hunk 2]
```diff
@@ -20,19 +20,21 @@
...
 
 public class Netty4HttpResponse extends DefaultFullHttpResponse implements HttpResponse {
 
-    private final Netty4HttpRequest request;
+    private final HttpHeaders requestHeaders;
 
-    Netty4HttpResponse(Netty4HttpRequest request, RestStatus status, BytesReference content) {
-        super(request.nettyRequest().protocolVersion(), HttpResponseStatus.valueOf(status.getStatus()), Netty4Utils.toByteBuf(content));
-        this.request = request;
+    Netty4HttpResponse(HttpHeaders requestHeaders, HttpVersion version, RestStatus status, BytesReference content) {
```
Comment (original-brownbear) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpResponse.java:35]: I figured if we release the request before using these parts (version and headers) it's a little nicer not to pass the full request around. Also, for non-pooled allocation (this includes all requests that aren't search or bulk) this makes the GC's life a little easier)

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/cors/Netty4CorsHandler.java [hunk 3]
```diff
@@ -93,20 +93,20 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
     public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
         assert msg instanceof Netty4HttpResponse : "Invalid message type: " + msg.getClass();
         Netty4HttpResponse response = (Netty4HttpResponse) msg;
-        setCorsResponseHeaders(response.getRequest().nettyRequest(), response, config);
+        setCorsResponseHeaders(response.requestHeaders(), response, config);
         ctx.write(response, promise);
     }
 
-    public static void setCorsResponseHeaders(HttpRequest request, HttpResponse resp, CorsHandler.Config config) {
+    public static void setCorsResponseHeaders(HttpHeaders headers, HttpResponse resp, CorsHandler.Config config) {
```
Comment (ywelsch) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/cors/Netty4CorsHandler.java:100]: Do the same for NioCorsHandler here?

**Review by ywelsch — CHANGES_REQUESTED** [review]
Let's align NIO in this PR as well (same transformation)

**Comment by original-brownbear** [comment]
Thanks Yannick, aligned the NIO implementation in 3b7ebcf

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Yannick!

**Pull request merged**